### ISSUE TITLE
Clone always fails on Windows 10

### ISSUE THREAD

**Issue by paulhiggs** [github/VisualStudio#948]
I have Visual Studio 2017 and the GitHib extension installed on Windows 7 and Windows 10 computers. Windows 7 runs perfectly. 
On Windows 10, I always receive the error message...

> Error encountered while cloning the remote repository: Git failed with a fatal error.
> CloneCommand.ExecuteClone

When trying to clone a repository. It doesnt matter which directory I choose to clone into. Doing a maunal clone from the Windows command line works perfectly --> Git command line is installed OK and permissions are OK

**Comment by paulhiggs** [comment]
extension.log contains this...
`2017-04-04 05:11:54.4898|ERROR|thread: 1|RepositoryCloneService|Could not clone https://github.com/paulhiggs/ProxyCtl.git to C:\Users\paul_\Source\Repos\ProxyCtl. Microsoft.TeamFoundation.Git.Api.Cli.GitFatalException: Git failed with a fatal error.
CloneCommand.ExecuteClone
   at Microsoft.TeamFoundation.Git.Api.Cli.GitCommand.TestExitCode(Int32 exitCode, String command, String errorText, Int32[] safeCodes)
   at Microsoft.TeamFoundation.Git.Api.Cli.CloneCommand.ExecuteClone(StringBuffer command, OperationProgressDelegate progressCallback)
   at Microsoft.TeamFoundation.Git.Api.Cli.CloneCommand.NormalFromRemote(String remoteUrl, String localDirectory, CloneOptions options)
   at Microsoft.TeamFoundation.Git.Api.Repository.Clone(String remoteUrl, String localDirectory, CloneOptions options)
   at Microsoft.TeamFoundation.Git.CoreServices.SccCloneJob.Execute()
   at Microsoft.TeamFoundation.Git.CoreServices.SccVoidJob.ExecuteBase()
   at Microsoft.TeamFoundation.Git.CoreServices.SccJob.Execute(ISccServiceHost serviceHost, ProgressSink progressSink, ProgressLabelSink progressLabelSink)
   at Microsoft.TeamFoundation.Git.CoreServices.SccJobService.RunSccJob(Object state)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.TeamFoundation.Git.Controls.SccUIService.<CloneAsync>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.Services.VSGitServices.<>c__DisplayClass5_0.<<Clone>b__0>d.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 84
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.VisualStudio.Threading.JoinableTask.<JoinAsync>d__76.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.Services.VSGitServices.<Clone>d__5.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 81
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitHub.Services.RepositoryCloneService.<CloneRepository>d__6.MoveNext() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.App\Services\RepositoryCloneService.cs:line 65`

GitHubVisualStudio-te.log contains
`Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
Error loading the repository list from the registry 'System.NullReferenceException: Object reference not set to an instance of an object.
   at GitHub.TeamFoundation.RegistryHelper.PokeTheRegistryForRepositoryList() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.15\RegistryHelper.cs:line 26
   at GitHub.Services.VSGitServices.GetKnownRepositories() in Y:\jenkins\workspace\GitHub for Visual Studio speakeasy jenkinskeep\src\GitHub.TeamFoundation.14\Services\VSGitServices.cs:line 118'
`

**Comment by shana** [comment]
Apologies for the delay in responding to your issue, it fell through the cracks somehow 😕 

Is this still happening to you? This looks to be a VS error - we use VS to clone and don't do it ourselves. If this is still happening, could you try copy-pasting the URL directly into the Local Git Repositories Clone area in Team Explorer and seeing if that works?

**Comment by jcansdale** [comment]
@paulhiggs,

From the logs, it looks like the following registry key doesn't exist:
`HKEY_CURRENT_USER\Software\Microsoft\VisualStudio\14.0\TeamFoundation\GitSourceControl\Repositories`

I'm wondering if the follwing key might not exist either:
HKEY_CURRENT_USER\Software\Microsoft\VisualStudio\14.0\TeamFoundation

Could you have a look and let me know what's there?

**Comment by paulhiggs** [comment]
The key  `HKEY_CURRENT_USER\Software\Microsoft\VisualStudio\14.0\TeamFoundation\GitSourceControl\Repositories` does exist on my computer (see "snip1.png") however it does not exist under `HKEY_CURRENT_USER\Software\Microsoft\VisualStudio\15.0\TeamFoundation\` (see "snip2.png")
![snip2](https://cloud.githubusercontent.com/assets/3158807/25537721/aa6b7b92-2c0d-11e7-9120-afe1d69106e8.PNG)
![snip1](https://cloud.githubusercontent.com/assets/3158807/25537720/aa6b0ce8-2c0d-11e7-873b-11bde8c9e16d.PNG)

**Comment by shana** [comment]
```
2017-04-04 05:11:54.4898|ERROR|thread: 1|RepositoryCloneService|Could not clone https://github.com/paulhiggs/ProxyCtl.git to C:\Users\paul_\Source\Repos\ProxyCtl. Microsoft.TeamFoundation.Git.Api.Cli.GitFatalException: Git failed with a fatal error.
CloneCommand.ExecuteClone
   at Microsoft.TeamFoundation.Git.Api.Cli.GitCommand.TestExitCode(Int32 exitCode, String command, String errorText, Int32[] safeCodes)
   at Microsoft.TeamFoundation.Git.Api.Cli.CloneCommand.ExecuteClone(StringBuffer command, OperationProgressDelegate progressCallback)
   at Microsoft.TeamFoundation.Git.Api.Cli.CloneCommand.NormalFromRemote(String remoteUrl, String localDirectory, CloneOptions options)
   at Microsoft.TeamFoundation.Git.Api.Repository.Clone(String remoteUrl, String localDirectory, CloneOptions options)
   at Microsoft.TeamFoundation.Git.CoreServices.SccCloneJob.Execute()
   at Microsoft.TeamFoundation.Git.CoreServices.SccVoidJob.ExecuteBase()
   at Microsoft.TeamFoundation.Git.CoreServices.SccJob.Execute(ISccServiceHost serviceHost, ProgressSink progressSink, ProgressLabelSink progressLabelSink)
   at Microsoft.TeamFoundation.Git.CoreServices.SccJobService.RunSccJob(Object state)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
```

@aarnott This looks to be a failure in the TE Clone API but can't tell from the log what the actual failure is. Is there a way of getting more information on the `Git fatal error`?

**Comment by paulhiggs** [comment]
There is no other information about the GIT fatal error provided in Visual Studio.
I can do a `git clone https://github.com/paulhiggs/ProxyCtl.git` to a temporary directory from a command prompt on the same computer and it clones OK

**Comment by paulhiggs** [comment]
I just tried again and see that the target folder for the clone is created, just no files cloned into it.

**Comment by paulhiggs** [comment]
@shana, previous;y you wrote..
> Is this still happening to you? This looks to be a VS error - we use VS to clone and don't do it ourselves. If this is still happening, could you try copy-pasting the URL directly into the Local Git Repositories Clone area in Team Explorer and seeing if that works?

Yes, this is still happening. I tried to clone from the Local Git Repositories Clone area and get the same error.
![image](https://cloud.githubusercontent.com/assets/3158807/25711696/933a2b4e-30bd-11e7-8480-f41356d63221.png)

**Comment by AArnott** [comment]
@shana I've asked a TE team member to respond.

**Comment by whoisj** [comment]
Thanks @AArnott. It is difficult to tell from collected data what is happening / failing.

@paulhiggs could you set a couple of environment variables, to collection additional logs?

 1. Open Command Prompt
 2. Type `setx GIT_TRACE %userprofile%/git.log`
 3. Type `setx GCM_TRACE %userprofile%/git.log`
 4. Re-open Visual Studio
 5. Attempt the clone, reproduce the failure.
 6. Copy the contents of `%userprofile%\git.log` and paste them as a response to this thread.

The log file will grow quickly, so disabling the logging is likely a good idea. I never disable mine, but I also have a scheduled task (at logon) to delete the log file. Disabling the logging is done by removing the environment variable.

 1. Open Command Prompt
 2. Type `setx GIT_TRACE ""`
 3. Type `setx GCM_TRACE ""`

Thanks - I'll take a look at the information asap. As a note, any issues you have with Visual Studio are likely best reported on https://developercommunity.visualstudio.com . We have an army of program managers trolling (as in "fish by trailing a baited line along behind a boat.") the site for issues to be fixed.

**Comment by paulhiggs** [comment]
@whoisj the log only contained two lines (with the clone failing)
`12:45:08.726734 git.c:371               trace: built-in: git 'clone' '--progress' '--verbose' '--recurse-submodules' 'https://github.com/paulhiggs/ProxyCtl.git' 'C:\Users\paul_\Source\Repos\ProxyCtl' `
`12:45:08.833022 run-command.c:350       trace: run_command: 'git-remote-https' 'origin' 'https://github.com/paulhiggs/ProxyCtl.git' `

**Comment by whoisj** [comment]
@paulhiggs any chance that there's a http proxy at play here? 

Also, I expected to see logging from the GCM. If it is truly not logging, then `git-remote-https` is failing early - I've only seen that when there's anti-malware that has falsely identified Git as a threat.

**Comment by paulhiggs** [comment]
@whoisj no proxy in the path!
This works....
`C:\>git clone https://github.com/paulhiggs/ProxyCtl.git %userprofile%\Source\Repos\ProxyCtl`
`Cloning into 'C:\Users\paul_\Source\Repos\ProxyCtl'... `
`remote: Counting objects: 18, done. `
`remote: Total 18 (delta 0), reused 0 (delta 0), pack-reused 18 `
`Unpacking objects: 100% (18/18), done.`

**Comment by shanlun** [comment]
I am observing this same problem, fresh Windows 10 install on with VS2017 and GH extension 2.4.3.1737.

**Comment by grokys** [comment]
@whoisj what is the status on this now? It seems to be the same problem as reported here: https://developercommunity.visualstudio.com/content/problem/19752/git-cant-clone-remote-repository.html

**Comment by whoisj** [comment]
@grokys the issue linked has already been resolved and shipped in VS2017; v15.3 I think. 

There are however, a number of anti-malware systems which prevent git.exe from running outside of the expected `%ProgramFiles%\Git` directory. Adding exclusions for the `git.exe` found under `%ProgramFiles%\Microsoft Visual Studio` directory often alleviates this issue.
